feat(controller): add incremental namespace sync feature

tricktron opened the pull request:
## Incremental Namespace Sync - Alpha Feature

Closes #23806

This PR introduces **incremental namespace synchronization** for Argo CD clusters, optimizing cluster cache updates when namespaces are added or removed from cluster configurations. It is implemented behind a feature flag and stays fully-backwards compatible.

### Problem Statement

Currently, when Argo CD detects namespace changes in a cluster configuration, it invalidates the entire cluster cache and performs a full resync of all namespaces. For **namespace-mode** clusters with many namespaces (100+), this can be slow and resource-intensive:

- Requires listing all resources across all namespaces
- Re-establishes all watches (one per namespace per API)
- Blocks application reconciliation during the entire sync operation
- Generates significant API server load

I have added a [script](https://github.com/tricktron/argo-cd/blob/24562a21ac0bbf69b5d07c3f44c673d0b822e5d6/benchmark/README.md) to easily reproduce the problem and test this feature in a local Kind cluster.

### Solution

This PR implements incremental namespace sync using a diff-based approach:

1. **Namespace change detection** - Compare desired vs current namespace sets
2. **Targeted updates** - Only sync added/removed namespaces
3. **Cache preservation** - Keep existing namespace caches intact
4. **Graceful fallback** - Revert to full invalidation on errors

### Implementation Details

The main idea is that we Introduces a new namespace specific watch and cancel data structure where the namespace-specific contexts depend on the existing global API resource contexts so that the full cluster invalidation API context cancellation propagates to the namespace-specific contexts.

**Gitops-Engine Layer** (`gitops-engine/pkg/cache/cluster.go`):
- `AddNamespace(namespace string) error` - Incrementally adds namespace and syncs its resources
- `RemoveNamespace(namespace string) error` - Incrementally removes namespace and its resources
- `WithIncrementalNamespaceSync(enabled bool)` - Feature flag option
- Per-namespace watch contexts for selective cancellation

**Controller Layer** (`controller/cache/cache.go`):
- `handleNamespaceChanges()` - Detects namespace diffs and calls engine APIs
- Feature flag: `ARGOCD_ENABLE_INCREMENTAL_NAMESPACE_SYNC` (opt-in)
- CLI flag: `--incremental-namespace-sync-enabled`
- Automatic fallback to full invalidation on errors

### Testing

**Unit tests** (`gitops-engine/pkg/cache/cluster_incremental_sync_test.go`):
- ✅ AddNamespace with feature enabled/disabled
- ✅ RemoveNamespace with feature enabled/disabled
- ✅ Resource syncing and watch management
- ✅ Error handling and fallback behavior
- ✅ Race condition detection

**Integration tests** (`controller/cache/cache_test.go`):
- ✅ Namespace additions with incremental sync
- ✅ Namespace removals with incremental sync
- ✅ Simultaneous add/remove operations
- ✅ Error handling with fallback
- ✅ Backward compatibility (feature disabled)

**E2E tests** (`test/e2e/cluster_test.go`):
- ✅ `TestClusterIncrementalAddNamespace` - Multi-app cluster with namespace addition
- ✅ `TestClusterIncrementalRemoveNamespace` - Multi-app cluster with namespace removal

### Documentation

- ✅ Operator manual: `docs/operator-manual/incremental-namespace-sync.md`

Checklist:

* [] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Title of the PR](https://argo-cd.readthedocs.io/en/latest/developer-guide/submit-your-pr/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

tricktron committed:
feat(cache): support adding namespaces incrementally to cluster cache

Enable cluster cache to watch additional namespaces at runtime without
requiring a full cache rebuild. This implements the core mechanism for
incremental namespace synchronization.

Implementation details:
- Check feature flag before allowing namespace additions
- Start watches for all resource types in the new namespace
- Sync existing resources from the added namespace
- Propagate errors from watch initialization

Signed-off-by: Thibault Gagnaux <thibault.gagnaux@bit.admin.ch>

tricktron committed:
feat(cache): support removing namespaces incrementally from cluster cache

Enable cluster cache to stop watching namespaces at runtime and clean
up associated resources. This complements dynamic namespace addition
for incremental namespace synchronization.

Implementation details:
- Check feature flag before allowing namespace removal
- Stop all watches for resources in the removed namespace
- Remove cached resources from the removed namespace
- Clean up namespace tracking data structures

Signed-off-by: Thibault Gagnaux <thibault.gagnaux@bit.admin.ch>

tricktron committed:
refactor(cache): track namespace state with new data structure

Introduces a new namespace specific watch and cancel data structure
where the namespace-specific contexts depend on the existing global API
resource contexts so that the full cluster invalidation API context
cancellation propagates to the namespace-specific contexts.

Besides, this restructures the gitops-engine cache tests to improve
clarity and resolve a flaky test caused by improper namespace state
tracking.

Signed-off-by: Thibault Gagnaux <thibault.gagnaux@bit.admin.ch>

tricktron committed:
chore(cache): regenerate mocks for cluster cache interface

Update generated mocks to include new methods for incremental namespace
management in the cluster cache interface.

Signed-off-by: Thibault Gagnaux <thibault.gagnaux@bit.admin.ch>

tricktron committed:
feat(controller): add incremental namespace sync feature flag

Connect the incremental namespace sync mechanism to the application
controller's cache, allowing the controller to incrementally update the
cluster cache instead of performing full cache rebuilds when the feature
is enabled.

Implementation details:
- Add ARGOCD_ENABLE_PARTIAL_NAMESPACE_CLUSTER_SYNC feature flag
- Detect namespace additions and removals by comparing tracked namespaces
- Delegate namespace changes to gitops-engine cluster cache
- Extract namespace change handling into dedicated function for clarity

Signed-off-by: Thibault Gagnaux <thibault.gagnaux@bit.admin.ch>

tricktron committed:
test(controller): expand test coverage for namespace change detection

Add comprehensive tests for the controller's ability to detect and
handle namespace changes during cluster cache synchronization.

Signed-off-by: Thibault Gagnaux <thibault.gagnaux@bit.admin.ch>

tricktron committed:
test: add e2e tests for incremental namespace sync

Add end-to-end tests verifying that applications correctly sync when
namespaces are added or removed with the incremental sync feature enabled.

Signed-off-by: Thibault Gagnaux <thibault.gagnaux@bit.admin.ch>

tricktron committed:
chore(cache): add logging for namespace operations

Add structured logs for namespace addition and removal operations
to aid troubleshooting and monitoring of incremental namespace sync.

Signed-off-by: Thibault Gagnaux <thibault.gagnaux@bit.admin.ch>

tricktron committed:
docs: add incremental namespace sync feature documentation

Signed-off-by: Thibault Gagnaux <thibault.gagnaux@bit.admin.ch>

anandf commented on controller/cache/cache.go:
```diff
@@ -824,6 +829,79 @@ func (c *liveStateCache) handleAddEvent(cluster *appv1.Cluster) {
+	cluster clustercache.ClusterCache,
+	oldNamespaces, newNamespaces []string,
+) []clustercache.UpdateSettingsFunc {
+	if reflect.DeepEqual(oldNamespaces, newNamespaces) {
```
`reflect.DeepEqual` will return false if the same namespaces are providedd, but in different order. It is better to use semantic equality check, Consider using `k8s.io/apimachinery/pkg/api/equality.Semantic.DeepEqual` https://pkg.go.dev/k8s.io/apimachinery/pkg/api/equality

anandf reviewed (changes requested):
Thanks @tricktron for submitting this change. I have posted few minor comments.

anandf reviewed (approved):
LGTM

jannfis reviewed (commented):
Thanks for your PR. I have some general remarks, please see below.

Also, is there a reason you chose to hide this behind a feature flag?

tricktron committed:
fix(cache): prevent nil panic in type assertion error handling

Signed-off-by: Thibault Gagnaux <thibault.gagnaux@bit.admin.ch>

tricktron committed:
fix(cache): add respect respect rbac strict check for namespace

Differentiate between RespectRbacNormal and RespectRbacStrict by
additionally verifying self subject access review for RespectRbacStrict
mode.

Signed-off-by: Thibault Gagnaux <thibault.gagnaux@bit.admin.ch>

tricktron committed:
refactor(cache): remove named return and move logging to caller

Signed-off-by: Thibault Gagnaux <thibault.gagnaux@bit.admin.ch>

tricktron commented on gitops-engine/pkg/cache/cluster.go:
```diff
@@ -490,6 +503,162 @@ func (c *clusterCache) Invalidate(opts ...UpdateSettingsFunc) {
+			un, ok := obj.(*unstructured.Unstructured)
+			if !ok {
+				return fmt.Errorf("object %s/%s has an unexpected type", un.GroupVersionKind().String(), un.GetName())
+			}
```
@jannfis Good catch. I blindly copied that from the existing source code. I fixed in now in all three occurrences.

tricktron commented:
> Thanks for your PR. I have some general remarks, please see below.
> 
> Also, is there a reason you chose to hide this behind a feature flag?

@jannfis Thank you for your great review comments. This feature is mainly a performance improvement for users who install and use ArgoCD in namespaced-mode. Rather than possibly breaking all users, I designed it behind a feature-flag so that affected users can opt-in to test it. Besides, I wanted to limit the blast radius because the gitops-engine code is at the core of the sync functionality.

tricktron committed:
test: fix cluster E2E rebase API mismatches

Align cluster e2e tests with the context-based fixture API after rebase.

Signed-off-by: Thibault Gagnaux <thibault.gagnaux@bit.admin.ch>

tricktron committed:
test: fix cluster fixture suffixing "in-cluster" name

The test isolation refactor missed guarding well-known cluster
names in the cluster fixture, causing SetNamespaces and GetByName
to target a non-existent cluster entry.

Signed-off-by: Thibault Gagnaux <thibault.gagnaux@bit.admin.ch>

tricktron committed:
fix: use fork module path for gitops-engine imports

CI resolves github.com/argoproj/gitops-engine to upstream v0.7.3
which is incompatible with Go 1.26 kubectl APIs. Use the fork
path github.com/argoproj/argo-cd/gitops-engine everywhere.

Signed-off-by: Thibault Gagnaux <thibault.gagnaux@bit.admin.ch>

agaudreault commented on docs/operator-manual/incremental-namespace-sync.md:
````diff
@@ -0,0 +1,137 @@
+
+Restart the controller:
+
+```bash
+kubectl rollout restart statefulset argocd-application-controller -n argocd
+```
+
+### Method 2: Via Environment Variable
+
+Edit the controller StatefulSet:
+
+```bash
+kubectl edit statefulset argocd-application-controller -n argocd
+```
+
+Add the environment variable:
+
+```yaml
+spec:
+  template:
+    spec:
+      containers:
+      - name: argocd-application-controller
+        env:
+        - name: ARGOCD_ENABLE_INCREMENTAL_NAMESPACE_SYNC
+          value: "true"
+```
+
+The controller pods will automatically restart to apply the change.
+
+### Method 3: Using kubectl set env
+
+Set the environment variable directly:
+
+```bash
+kubectl set env statefulset/argocd-application-controller \
+  ARGOCD_ENABLE_INCREMENTAL_NAMESPACE_SYNC=true -n argocd
+```
+
+This will trigger a rolling update of the controller pods.
````
Let's keep it GitOps and simple



### Method 1: Via ConfigMap (Recommended)

Update the `argocd-cmd-params-cm` ConfigMap:

```bash
apiVersion: v1
kind: ConfigMap
metadata:
  name: argocd-cmd-params-cm
data:
  controller.incremental.namespace.sync.enabled: "true"
```

**Note**: You must restart the controller pods for the setting to apply.

### Method 2: Via Environment Variable

Add the environment variable:

```yaml
spec:
  template:
    spec:
      containers:
      - name: argocd-application-controller
        env:
        - name: ARGOCD_ENABLE_INCREMENTAL_NAMESPACE_SYNC
          value: "true"
```

agaudreault commented on test/e2e/fixture/context.go:
```diff
@@ -83,6 +83,12 @@ func (s *TestState) SetName(name string) {
 	s.name = DnsFriendly(strings.TrimSuffix(name, suffix), suffix)
 }
 
+// SetNameRaw sets the name without appending a suffix.
+// Use this for well-known names (e.g. "in-cluster") that must not be modified.
+func (s *TestState) SetNameRaw(name string) {
+	s.name = name
+}
+
```
This cannot be done, as tests must use unique resource to prevent collision when running in parallel.

agaudreault commented on test/e2e/cluster_test.go:
```diff
@@ -167,6 +169,139 @@ func TestClusterSet(t *testing.T) {
+	testCtx := fixture.EnsureCleanState(t)
+	existingNs := testCtx.DeploymentNamespace()
+	newNs := "new-ns"
+	t.Setenv("ARGOCD_ENABLE_INCREMENTAL_NAMESPACE_SYNC", "true")
```
This sets a variable on the test, not on the controller pods. You will need to update the configMap settings. And the e2e framework does not have the capability to restart a pod, so to implement this test, you will need the setting to apply without restart.

agaudreault commented on test/e2e/cluster_test.go:
```diff
@@ -167,6 +169,139 @@ func TestClusterSet(t *testing.T) {
+
+	// When: A new namespace is created and added to the cluster and a new
+	// app is deployed to it
+	errors.NewHandler(t).FailOnErr(fixture.Run("", "kubectl", "create", "namespace", newNs))
```
There is a convention to create namespaces with a specific label `"e2e.argoproj.io"` so they get automatically cleaned up. The namespace should also always be created with a unique suffix so it is unique to that test execution.

agaudreault commented on test/e2e/cluster_test.go:
```diff
@@ -167,6 +169,139 @@ func TestClusterSet(t *testing.T) {
+		Sync()
+
+	// Then: both the new app and the existing app are healthy and synced
+	GivenWithSameState(testCtx).
```
do not reuse GivenWithSameState. Instead, save the context in a variable
```
app2Expectations := GivenWithSameState(testCtx).
		Name("app-2").
		Path(guestbookPath).
		When().
		CreateWithNoNameSpace("--dest-namespace", newNs).
		Sync().
        Then()

app2Expectations.
		Expect(OperationPhaseIs(OperationSucceeded)).
		Expect(SyncStatusIs(SyncStatusCodeSynced)).
		Expect(HealthIs(health.HealthStatusHealthy))

```

agaudreault commented on controller/cache/cache.go:
```diff
@@ -228,6 +230,8 @@ type liveStateCache struct {
 	clusterSharding      sharding.ClusterShardingCache
 	resourceTracking     argo.ResourceTracking
 	ignoreNormalizerOpts normalizers.IgnoreNormalizerOpts
+	// enableIncrementalNamespaceSync configures all cluster caches with incremental namespace sync
+	enableIncrementalNamespaceSync bool
```
Instead of having a parameter set as a command argument, use the argocd-cm and load it from the configMap with the `settingsMgr`. This enables changing the setting without reloading the pod. Take a look at `cacheSettings` and `loadCacheSettings` for examples.

agaudreault commented on controller/cache/cache.go:
```diff
@@ -845,6 +850,78 @@ func (c *liveStateCache) handleAddEvent(cluster *appv1.Cluster) {
+	for _, ns := range added {
+		if err := cluster.AddNamespace(ns); err != nil {
+			log.Warnf("Failed to incrementally add namespace %s for cluster %s: %v", ns, cluster.GetClusterInfo().Server, err)
+			hasErrors = true
```
Should break on error since we will do the full SetNamespaces() and invalidation later anyway

agaudreault commented on gitops-engine/pkg/cache/cluster.go:
```diff
@@ -625,6 +638,170 @@ func (c *clusterCache) Invalidate(opts ...UpdateSettingsFunc) {
+// otherwise falls back to full cluster cache invalidation
+func (c *clusterCache) AddNamespace(namespace string) error {
+	if !c.incrementalNamespaceSync {
+		c.Invalidate(func(cache *clusterCache) {
+			cache.namespaces = append(cache.namespaces, namespace)
+		})
```
Calling invalidate here without knowing if the Caller will call EnsureSynced is problematic. 

Instead, return an error stating that this cannot be used if incrementalNamespaceSync is not enabled. Or remove the flag condition from the cache, and only use the feature flag in the controller to use or not the feature

agaudreault commented on gitops-engine/pkg/cache/cluster.go:
```diff
@@ -625,6 +638,170 @@ func (c *clusterCache) Invalidate(opts ...UpdateSettingsFunc) {
+	apisToSync := c.snapshotApisMeta()
+	c.lock.Unlock()
+
+	err := c.syncNamespaceResources(namespace, apisToSync)
```
This will not happen in the cluster cache lock, which will be problematic with any fields manipulation of the cluster cache, including setting the namespaceCancels.

Since the caller of sync() must have the whole cluster lock, I recommend having the whole cluster lock as well to sync one namespace.

agaudreault commented on gitops-engine/pkg/cache/cluster.go:
```diff
@@ -625,6 +638,170 @@ func (c *clusterCache) Invalidate(opts ...UpdateSettingsFunc) {
+	}
+
+	// Feature enabled: incremental sync
+	c.log.Info("Start syncing namespace", "namespace", namespace)
```
Add log after taking the lock

agaudreault commented on gitops-engine/pkg/cache/cluster.go:
```diff
@@ -625,6 +638,170 @@ func (c *clusterCache) Invalidate(opts ...UpdateSettingsFunc) {
+		}
+	}
+
+	c.lock.Unlock()
```
use defer after taking the lock

agaudreault commented on gitops-engine/pkg/cache/cluster.go:
```diff
@@ -625,6 +638,170 @@ func (c *clusterCache) Invalidate(opts ...UpdateSettingsFunc) {
+	return result
+}
+
+func (c *clusterCache) snapshotApisMeta() map[schema.GroupKind]*apiMeta {
+	snapshot := make(map[schema.GroupKind]*apiMeta)
+	maps.Copy(snapshot, c.apisMeta)
+	return snapshot
+}
```
Do not create copy since modifying the value of apiMeta object is an operation that can cause race conditions.

agaudreault commented on gitops-engine/pkg/cache/cluster.go:
```diff
@@ -625,6 +638,170 @@ func (c *clusterCache) Invalidate(opts ...UpdateSettingsFunc) {
+	c.log.Info("Start syncing namespace", "namespace", namespace)
+
+	c.lock.Lock()
+	c.namespaces = append(c.namespaces, namespace)
```
What if the namespace is already in the list? This function should make sure not to add it twice

agaudreault commented on gitops-engine/pkg/cache/cluster.go:
```diff
@@ -625,6 +638,170 @@ func (c *clusterCache) Invalidate(opts ...UpdateSettingsFunc) {
+				return fmt.Errorf("object has unexpected type: %T", obj)
+			}
+			newRes := c.newResource(un)
+			c.lock.Lock()
```
No need for this if the parent lock is held by the caller

agaudreault reviewed (changes requested):
The feature looks good to me and is well documented. I have some concern that this does not fully fix the problem, since the cache will still be invalidated every now and then, causing the Apps to go in that stuck refreshing state. This is most likely caused by lock contention, usually a combination of the githops-engine locks, and the K8S API client configurations. 

I think dynamically adjusting these locks to remove the contention is a feature that would be necessary to fix the problem as a whole. This feature still have value since it reduces the amount of time watches are invalidated.

- Setting should be able to be changed without restarts.
- Locking should be consistent with existing behaviour
- The e2e tests need to be reworked. I doubt they are actually validating the feature, and I don't think they will fail if the feature is not behaving correctly.

Mangaal commented:
Hi @tricktron, thanks for all the effort you've put into this feature! Just wanted to see if you are you still working on this PR and planning to address the recent review.

uittenbroekrobbert commented:
@tricktron thanks for working on this, we're really hoping it lands. Without it
our production stalls for minutes on every namespace creation. We use ArgoCD
namespace-scoped within a tenant cluster.

On the open lock comment: we tried to reproduce it and ended up with a patch
that might help. It adds a small concurrent `-race` test and a fix, and the
whole cache package passes under `-race` with it. Totally your call whether it
fits, happy to adjust. If it's useful it should apply with `git am`.

I'd also gladly help with the argocd-cm config reload item or any of the other open
comments. Just let me know.

```
From 8c962e1d375fc54ae7cccda3a28c55a773a1d998 Mon Sep 17 00:00:00 2001
From: Robbert Uittenbroek <robbert.uittenbroek@rijksoverheid.nl>
Date: Mon, 29 Jun 2026 14:22:44 +0200
Subject: [PATCH] fix(cache): lock namespaceCancels write in incremental
 namespace sync

syncNamespaceResources wrote the shared apiMeta.namespaceCancels map
outside c.lock. snapshotApisMeta copies the map but shares the *apiMeta
pointers, so concurrent AddNamespace calls (and RemoveNamespace, which
holds c.lock) wrote the same Go map at once, tripping the race detector
and the 'concurrent map writes' check.

Lock only around the map write; the namespace list stays unlocked so I/O
is not serialized. Adds a concurrent -race test.

Signed-off-by: Robbert Uittenbroek <robbert.uittenbroek@rijksoverheid.nl>
---
 gitops-engine/pkg/cache/cluster.go            |  3 +
 .../cluster_incremental_concurrent_test.go    | 56 +++++++++++++++++++
 2 files changed, 59 insertions(+)
 create mode 100644 gitops-engine/pkg/cache/cluster_incremental_concurrent_test.go

diff --git a/gitops-engine/pkg/cache/cluster.go b/gitops-engine/pkg/cache/cluster.go
index 2a99c5e01..1dc41dc42 100644
--- a/gitops-engine/pkg/cache/cluster.go
+++ b/gitops-engine/pkg/cache/cluster.go
@@ -747,7 +747,10 @@ func (c *clusterCache) syncNamespaceResources(namespace string, apisToSync map[s
 
 		// Create per-namespace context derived from the API's parent context
 		nsCtx, nsCancel := context.WithCancel(meta.watchCtx)
+		// namespaceCancels is shared across goroutines, lock the write
+		c.lock.Lock()
 		meta.namespaceCancels[namespace] = nsCancel
+		c.lock.Unlock()
 
 		if err := c.loadNamespaceResources(nsCtx, client, clientset, api, namespace); err != nil {
 			return fmt.Errorf("failed to load resources for %s in namespace %s: %w", gk.String(), namespace, err)
diff --git a/gitops-engine/pkg/cache/cluster_incremental_concurrent_test.go b/gitops-engine/pkg/cache/cluster_incremental_concurrent_test.go
new file mode 100644
index 000000000..2c1e0b5e2
--- /dev/null
+++ b/gitops-engine/pkg/cache/cluster_incremental_concurrent_test.go
@@ -0,0 +1,56 @@
+package cache
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"k8s.io/client-go/rest"
+)
+
+// TestAddNamespaceConcurrent runs concurrent AddNamespace calls under -race to
+// guard against unsynchronized writes to the shared apiMeta.namespaceCancels map.
+func TestAddNamespaceConcurrent(t *testing.T) {
+	existingPod := createPod("existing-pod", "existing-namespace")
+	_, mockKubectl := setupFakeCluster(existingPod)
+	cache := NewClusterCache(
+		&rest.Config{},
+		SetKubectl(mockKubectl),
+		SetNamespaces([]string{"existing-namespace"}),
+		WithIncrementalNamespaceSync(true),
+	)
+	require.NoError(t, cache.EnsureSynced())
+
+	const goroutines = 16
+	var wg sync.WaitGroup
+	for i := 0; i < goroutines; i++ {
+		wg.Add(1)
+		ns := fmt.Sprintf("ns-%d", i)
+		go func() { defer wg.Done(); _ = cache.AddNamespace(ns) }()
+	}
+	wg.Wait()
+}
+
+// TestAddRemoveNamespaceConcurrent runs concurrent adds and removes under -race.
+func TestAddRemoveNamespaceConcurrent(t *testing.T) {
+	existingPod := createPod("existing-pod", "existing-namespace")
+	_, mockKubectl := setupFakeCluster(existingPod)
+	cache := NewClusterCache(
+		&rest.Config{},
+		SetKubectl(mockKubectl),
+		SetNamespaces([]string{"existing-namespace"}),
+		WithIncrementalNamespaceSync(true),
+	)
+	require.NoError(t, cache.EnsureSynced())
+
+	const goroutines = 16
+	var wg sync.WaitGroup
+	for i := 0; i < goroutines; i++ {
+		wg.Add(2)
+		ns := fmt.Sprintf("ns-%d", i)
+		go func() { defer wg.Done(); _ = cache.AddNamespace(ns) }()
+		go func() { defer wg.Done(); _ = cache.RemoveNamespace(ns) }()
+	}
+	wg.Wait()
+}
-- 
2.54.0

```

tricktron commented:
Hello @Mangaal and @uittenbroekrobbert, thank you for your feedback.

I paused my work on this bigger feature for the smaller related fix in https://github.com/argoproj/argo-cd/pull/27528. Would be nice if we could get this in first. I appreciate any help.
